### ISSUE TITLE
AaU, I want to configure the timeouts used by the AWSApiPlugin.

### ISSUE THREAD

**Issue by merveylcu** [aws-amplify/amplify-android#913]
I randomly get the following error when I send a request. (Regardless of GET, POST)
**`com.amplifyframework.api.ApiException: Received an IO exception while making the request.`**

Here's my code:

``` kotlin
fun <K, T> sendRequest(
        request: K? = null, api: APIs,
        queryMap: MutableMap<String, String>? = mutableMapOf(),
        pathParameters: MutableMap<String, String>? = mutableMapOf(),
        onSuccessAction: ((BaseResponse<T>) -> Unit)? = null
    ) {
        showIndicator()
        //Dynamic path parameters control
        var editedEndpoint = api.endpoint
        pathParameters?.forEach {
            editedEndpoint = editedEndpoint.replace("{${it.key}}", it.value)
        }

        when (api.method) {
            RequestMethod.POST -> {
                val json = gson.toJson(request)
                val bytes = json.toByteArray(StandardCharsets.UTF_8)
                
                val options = RestOptions.builder()
                    .addPath(editedEndpoint)
                    .addBody(bytes)
                    .addHeaders(
                        mutableMapOf(
                            "Content-Type" to "application/json",
                            "Accept" to "application/json"
                        )
                    )
                    .addQueryParameters(queryMap ?: mapOf())
                    .build()
                Amplify.API.post(options,
                    { response ->
                        when {
                            response.code.isSuccessful -> {

                                val hasData = response.data.asJSONObject().isNull("data").not()
                                if (hasData && api.responseType != Empty::class.java) {
                                    val typedData = gson.fromJson(
                                        response.data.asJSONObject()["data"].toString(),
                                        api.responseType
                                    )
                                    onSuccessAction?.invoke(BaseResponse(typedData) as BaseResponse<T>)
                                    hideIndicator()
                                } else {
                                    onSuccessAction?.invoke(BaseResponse(null))
                                    hideIndicator()
                                }
                            }
                            response.code.isClientError -> {
                                
                            }
                            response.code.isServiceFailure -> {
                                
                            }
                        }
                    },
                    {
                        //API Error
                    })
            }

            RequestMethod.GET -> {
                val options = RestOptions.builder()
                    .addPath(editedEndpoint)
                    .addHeaders(
                        mutableMapOf(
                            "Content-Type" to "application/x-www-form-urlencoded; charset=utf-8",
                            "Accept" to "application/json"
                        )
                    )
                    .addQueryParameters(queryMap ?: mapOf())
                    .build()
               
                Amplify.API.get(options,
                    { response ->
                        when {
                            response.code.isSuccessful -> {
                                val hasData = response.data.asJSONObject().isNull("data").not()
                                if (hasData && api.responseType != Empty::class.java) {
                                    val typedData = gson.fromJson(
                                        response.data.asJSONObject().optString("data"),
                                        api.responseType
                                    )
                                    onSuccessAction?.invoke(BaseResponse(typedData) as BaseResponse<T>)
                                    hideIndicator()
                                } else {
                                    onSuccessAction?.invoke(BaseResponse(null))
                                    hideIndicator()
                                }
                            }
                            response.code.isClientError -> {
                               
                            }
                            response.code.isServiceFailure -> {
                                
                            }
                        }
                    },
                    {
                        //API Error
                    })
            }
        }
    }
```

**Comment by jamesonwilliams** [comment]
Your snipped contains several references that are not part of our library. What part of it is triggering the exception? Where do you obtain the exception? Are you able to produce a smaller, more isolated code snippet, with clear reproduction steps?

**Comment by merveylcu** [comment]
> ```
> AmplifyException {message=Received an IO exception while making the request., cause=java.net.SocketTimeoutException: timeout, recoverySuggestion=Retry the request.}
>     at com.amplifyframework.api.aws.operation.AWSRestOperation$OkHttpCallback.onFailure(AWSRestOperation.java:131)
>     at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:525)
>     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
>     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
>     at java.lang.Thread.run(Thread.java:923)
> Caused by: java.net.SocketTimeoutException: timeout
>     at okhttp3.internal.http2.Http2Stream$StreamTimeout.newTimeoutException(Http2Stream.kt:677)
>     at okhttp3.internal.http2.Http2Stream$StreamTimeout.exitAndThrowIfTimedOut(Http2Stream.kt:686)
>     at okhttp3.internal.http2.Http2Stream.takeHeaders(Http2Stream.kt:143)
>     at okhttp3.internal.http2.Http2ExchangeCodec.readResponseHeaders(Http2ExchangeCodec.kt:96)
>     at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.kt:106)
>     at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:79)
>     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
>     at com.amplifyframework.api.aws.UserAgentInterceptor.intercept(UserAgentInterceptor.java:56)
>     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
>     at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34)
>     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
>     at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
>     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
>     at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
>     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
>     at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
>     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
>     at com.amplifyframework.api.aws.sigv4.AppSyncSigV4SignerInterceptor.intercept(AppSyncSigV4SignerInterceptor.java:236)
>     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
>     at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
>     at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:517)
> 	... 3 more
> 
> 
>  Error message = POST EXCEPTION Received an IO exception while making the request.
> ```

**Comment by merveylcu** [comment]
When I make a request, the API Error block is triggered before I get the answer.

Note: I'm waiting for a while when I make a request. Sometimes it gives this error after 5 seconds, sometimes 30 seconds.

Here's exception:
> ```
> AmplifyException {message=Received an IO exception while making the request., cause=java.net.SocketTimeoutException: timeout, recoverySuggestion=Retry the request.}
>     at com.amplifyframework.api.aws.operation.AWSRestOperation$OkHttpCallback.onFailure(AWSRestOperation.java:131)
>     at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:525)
>     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
>     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
>     at java.lang.Thread.run(Thread.java:923)
> Caused by: java.net.SocketTimeoutException: timeout
>     at okhttp3.internal.http2.Http2Stream$StreamTimeout.newTimeoutException(Http2Stream.kt:677)
>     at okhttp3.internal.http2.Http2Stream$StreamTimeout.exitAndThrowIfTimedOut(Http2Stream.kt:686)
>     at okhttp3.internal.http2.Http2Stream.takeHeaders(Http2Stream.kt:143)
>     at okhttp3.internal.http2.Http2ExchangeCodec.readResponseHeaders(Http2ExchangeCodec.kt:96)
>     at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.kt:106)
>     at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:79)
>     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
>     at com.amplifyframework.api.aws.UserAgentInterceptor.intercept(UserAgentInterceptor.java:56)
>     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
>     at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34)
>     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
>     at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
>     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
>     at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
>     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
>     at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
>     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
>     at com.amplifyframework.api.aws.sigv4.AppSyncSigV4SignerInterceptor.intercept(AppSyncSigV4SignerInterceptor.java:236)
>     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
>     at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
>     at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:517)
> 	... 3 more
> 
> 
>  Error message = POST EXCEPTION Received an IO exception while making the request.
> ```

**Comment by jamesonwilliams** [comment]
@merveylcu Are you intending to use HTTP/2? What is the endpoint?
> okhttp3.internal.http2.Http2Stream

Try disabling HTTP/2 and making a request to this endpoint via the escape hatch:
```kotlin
// Access the underlying OkHttpClient used by AWSApiPlugin
val apiName = "key from your amplifyconfiguration.json"
val plugin = Amplify.API.getPlugin("awsAPIPlugin") as AWSApiPlugin
val pluginClient = plugin.escapeHatch.get(apiName)

// Build a new one, that only allows HTTP/1.1.
val okhttp = pluginClient.newBuilder()
    .protocols(Arrays.asList(Protocol.HTTP_1_1))
    .build()

// Try making a get request with it.
val request = Request.Builder()
    .url("http://the.url.from.your.amplifyconfiguration.json")
    .build()
val response = okhttp.newCall(request).execute()
```

**Comment by merveylcu** [comment]
I got the error below when I tried it with the code you shared.

**`HTTP FAILED: java.net.SocketTimeoutException: timeout`**

<br/>

But when you added the code below to the code you shared, the whole problem was resolved.

```
val okHttp = pluginClient.newBuilder()
      .protocols(listOf(Protocol.HTTP_1_1))
      .connectTimeout(5, TimeUnit.MINUTES)
      .writeTimeout(5, TimeUnit.MINUTES)
      .readTimeout(5, TimeUnit.MINUTES)
      .build()
```

Is it possible to add timeout setting method to the amplify library? @jamesonwilliams 

Thank you for your help.

**Comment by aeraydogan** [comment]
For solving this situation, I just added below code in AWSApiPlugin class for configurate timeouts.

![image](https://user-images.githubusercontent.com/7622508/97785523-9c95e880-1bb6-11eb-9ef4-c35d387d64f9.png)

 I added new variables to ApiConfiguration class like

> private final int connectTimeout;    
> private final int writeTimeout;    
> private final int readTimeout;

And modified AWSApiPluginConfigurationReader according to these variables. At last I used new variables in our amplifyconfiguration.json.

> "api": {
>         "plugins": {
>             "awsAPIPlugin": {
>                 "OurApiKey": {
>                     "endpointType": "REST",
>                     "endpoint": "our_url",
>                     "region": "***",
>                     "authorizationType": "AWS_IAM",
**>                     "connectionTimeout": 300,
>                     "writeTimeout": 300,
>                     "readTimeout": 300**
>                 }
>             }
>         }
>     }

Can you add these configurations to AWSApiPlugin? Or is there any easy way to configurate timeouts?

**Comment by jamesonwilliams** [comment]
I agree this would be a good feature to include. Configurable timeouts is sort of a must-have for an API client. We will track enthusiasm according to user reactions on this ticket and will try to get it scheduled sooner or later. @aeraydogan Feel free to raise a PR! All in all we'd need:
1. Updates to the configuration objects and the `amplifyconfiguration.json` reader
2. The changes to the plugin that you propose
3. Updates on the docs site that talk about how to use it.

If the timeouts are not supplied in the `ampliflyconfiguration.json`, the current default values should be the same.

**Comment by SiddharthChoudhary** [comment]
Facing the exact same situation and error where I am trying to upload a 100MB sized file to AWS using AWSAmplify + REST Api solution. We really need a way to set connectionTimeout in this library

**Comment by SiddharthChoudhary** [comment]
Hey @jamesonwilliams,
So here's what I am trying to do. I am using AWS Amplify's Rest API solution to upload a file by converting it into byte streams and this goes to my lambda via AWS Amplify's REST API-> API Gateway-> Lambda. This lambda filters out and post the data in S3. It works fine if the file size is like 0-4.0 MB, but gives SocketConnectionTimeout error if greater than 6MB. 
I am trying to implement the approach you described above but the thing is making my custom REST API builder will need the AUTH Plugin as well, cause APIPlugin is dependent on AuthPlugin, even though I am not using it. Also, I was looking at a response code which I am getting through hitting the `Amplify.API.post(options,...)` and it's 413, which makes me wonder that if the API GW has a payload limit then would increasing connectionTimeout even work?
I am building this in Android.
Using: `com.amplifyframework:aws-api:1.16.11`

**Comment by jamesonwilliams** [comment]
I believe this issue will be resolved by #1207. To set custom timeouts, the user can do:
```java
AWSApiPlugin plugin = AWSApiPlugin.builder()
    .customizeClient("apiName", okHttpClientBuilder -> {
        okHttpClientBuilder.connectTimeout(10, TimeUnit.SECONDS);
    })
    .build();
Amplify.addPlugin(plugin);
```

**Comment by jamesonwilliams** [comment]
Resolved in [1.17.0](https://github.com/aws-amplify/amplify-android/releases/tag/release_v1.17.0).

**Comment by mfhenao88** [comment]
> I believe this issue will be resolved by #1207. To set custom timeouts, the user can do:
> 
> ```java
> AWSApiPlugin plugin = AWSApiPlugin.builder()
>     .customizeClient("apiName", okHttpClientBuilder -> {
>         okHttpClientBuilder.connectTimeout(10, TimeUnit.SECONDS);
>     })
>     .build();
> Amplify.addPlugin(plugin);
> ```

Hello @jamesonwilliams  I added the below code, but it isn't work  

`           AWSApiPlugin plugin = AWSApiPlugin.builder().configureClient(ApiName.clienteQiip.name(), okHttpClientBuilder -> {
                        okHttpClientBuilder
                                .protocols(Arrays.asList(Protocol.HTTP_1_1))
                                .connectTimeout(29, TimeUnit.SECONDS)
                                .writeTimeout(29, TimeUnit.SECONDS)
                                .readTimeout(29, TimeUnit.SECONDS)
                                .build();
                    })
                    .build();
            Amplify.addPlugin(plugin);

            Amplify.addPlugin(new AWSCognitoAuthPlugin());
            AmplifyConfiguration config = AmplifyConfiguration.builder(getApplicationContext()).devMenuEnabled(false).build();
            Amplify.configure(config, getApplicationContext());`
